File: ted-ext/ted-scheduler/src/test/java/ted/scheduler/I01SimpleTest.java
package ted.scheduler;

import org.junit.Before;
import org.junit.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import ted.driver.Ted.TedStatus;
import ted.driver.TedDriver;
import ted.driver.TedTask;
import ted.scheduler.TedSchedulerImpl.CronRetry;
import ted.scheduler.TedSchedulerImpl.Factory;
import ted.scheduler.utils.CronExpression;

import java.io.IOException;
import java.time.ZonedDateTime;
import java.util.Collections;
import java.util.Properties;

import static org.junit.Assert.assertEquals;
import static ted.scheduler.TestUtils.sleepMs;

public class I01SimpleTest {
	private final static Logger logger = LoggerFactory.getLogger(I01SimpleTest.class);

	private TedDriver driver;
	private TedScheduler scheduler;
	private TedSchedulerImpl schedulerImpl;
	private DaoPostgres dao;


	@Before
	public void init() throws IOException {
		Properties properties = TestUtils.readPropertiesFile("ted-I01.properties");
		driver = new TedDriver(TestConfig.testDbType, TestConfig.getDataSource(), properties);
		dao = new DaoPostgres(TestConfig.getDataSource(), properties.getProperty("ted.systemId"));
		driver.enablePrime();
		driver.start();
		scheduler = new TedScheduler(driver);
		schedulerImpl = new TedSchedulerImpl(driver);
		//this.context = driver.getContext();

	}

	@Test
	public void test1() {
		CronExpression cron = new CronExpression("0/11 14/19 0 1,11,21 1/2 *");
		ZonedDateTime zdt = cron.nextTimeAfter(ZonedDateTime.now());
		print((zdt = cron.nextTimeAfter(zdt)).toString());
		print((zdt = cron.nextTimeAfter(zdt)).toString());
		print((zdt = cron.nextTimeAfter(zdt)).toString());
	}

	@Test
	public void testSchd01() {
		scheduler.builder().name("TEST1")
				.scheduleCron("0 0/10 * 1/1 * *")
				.runnable(() -> {
					logger.info("executing scheduler task");
				}).register();
		sleepMs(2000);
	}

	@Test
	public void testMaint() {
		Long taskId = schedulerImpl.registerScheduler("TEST1", null,
				Factory.single(() -> logger.info("executing scheduler task")),
				new CronRetry("0 0/10 * ? * *"));
		print("scheduler taskId=" + taskId);
		TedTask task = driver.getTask(taskId);
		assertEquals(TedStatus.RETRY, task.getStatus());
		sleepMs(20);
		dao_execSql("update tedtask set status = 'ERROR', nextts = null where taskid = " + taskId + " returning taskid");
		sleepMs(20);
		task = driver.getTask(taskId);
		assertEquals(TedStatus.ERROR, task.getStatus());
		schedulerImpl.checkForErrorStatus();
		sleepMs(20);
		task = driver.getTask(taskId);
		assertEquals(TedStatus.RETRY, task.getStatus());
		sleepMs(2000);
	}


	public static void print(String msg){
		System.out.println(msg);
	}

	private void dao_execSql (String sql) {
		dao.selectData("test", sql, Void.class, Collections.emptyList());
	}

}

File: ted-ext/ted-scheduler/src/test/java/ted/scheduler/I01SimpleTest.kt
package ted.scheduler

import org.awaitility.Awaitility
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.slf4j.LoggerFactory
import ted.driver.Ted.TedStatus
import ted.driver.TedDriver
import ted.scheduler.impl.TedSchdDriverExt
import ted.scheduler.TestUtils.sleepMs
import ted.scheduler.impl.AbstractDao
import ted.scheduler.impl.TedSchedulerImpl
import ted.scheduler.impl.TedSchedulerImpl.*
import java.io.IOException
import java.util.concurrent.TimeUnit

class I01SimpleTest {
    private val logger = LoggerFactory.getLogger(I01SimpleTest::class.java)

    private lateinit var driver: TedDriver
    private lateinit var scheduler: TedScheduler
    private lateinit var schedulerImpl: TedSchedulerImpl
    private lateinit var systemId: String

    private lateinit var context: Context

    @Before
    @Throws(IOException::class)
    fun init() {
        val properties = TestUtils.readPropertiesFile("ted-I01.properties")
        driver = TedDriver(TestConfig.testDbType, TestConfig.getDataSource(), properties)
        driver.enablePrime()
        scheduler = TedScheduler(driver)
        schedulerImpl = TedSchedulerImpl(driver)

        systemId = properties.getProperty("ted.systemId")

        val tedSchdDriverExt = TedSchdDriverExt(driver)
        context = schedulerImpl.context
        context.thisSystem = tedSchdDriverExt.systemId()
        context.dataSource = tedSchdDriverExt.dataSource()
        context.dbType = tedSchdDriverExt.dbType()


        driver.start()
    }

    @After
    fun cleanup() {
        driver.shutdown()
        dao_execSql("update tedtask set nextts = null, status = 'DONE' where system = '$systemId' and "
            + " status in ('NEW', 'RETRY', 'WORK') returning taskid")
    }


    @Test
    fun `simple test scheduler every 1 second`() {
        var count = 0
        scheduler.builder().name("TEST1")
            .scheduleCron("* * * ? * *")
            .runnable {
                count++;
                logger.info("executing scheduler task $count")
            }
            .register()
        Awaitility.await().atMost(2500, TimeUnit.MILLISECONDS).until { count >= 2 }
    }

    @Test
    fun `should be only 1 task with same name`() {
        var count = 0
        var count2 = 0
        val shdId = scheduler.builder().name("TEST2")
            .scheduleCron("* * * ? * *")
            .runnable {
                count++;
                logger.info("[1] executing scheduler task $count")
            }
            .register()

        val shdId2 = scheduler.builder().name("TEST2")
            .scheduleCron("* * * ? * *")
            .runnable {
                count2++;
                logger.info("[2] executing scheduler task $count2")
            }
            .register()

        assertEquals(shdId, shdId2)

        // only 1st registration is active

        Awaitility.await().atMost(2500, TimeUnit.MILLISECONDS).until { count >= 2 }

        assertEquals(0, count2)
    }

    @Test
    fun `maintenance test`() {
        val taskId = schedulerImpl.registerScheduler("TEST1", null,
            Factory.single(Runnable { logger.info("executing scheduler task") }),
            CronRetry("0 0/10 * ? * *"))
        print("scheduler taskId=$taskId")
        sleepMs(20)

        var task = driver.getTask(taskId)!!
        assertTrue(setOf(TedStatus.RETRY, TedStatus.NEW).contains(task.status))

        sleepMs(20)
        dao_execSql("update tedtask set status = 'ERROR', nextts = null where taskid = $taskId returning taskid")
        sleepMs(20)

        task = driver.getTask(taskId)
        assertEquals(TedStatus.ERROR, task.status)
        schedulerImpl.checkForErrorStatus()

        sleepMs(20)

        task = driver.getTask(taskId)
        assertEquals(TedStatus.RETRY, task.status)

        sleepMs(2000)
    }

    private fun dao_execSql(sql: String) {
        (context.dao as AbstractDao).selectData("test", sql, Void::class, listOf())
    }

    fun print(msg: String) {
        println(msg)
    }

}
